pkg/executor/aggfuncs/spill_serialize_helper.go

```diff
@@ -0,0 +1,264 @@
+// limitations under the License.
+
+package aggfuncs
+
+import (
+	"unsafe"
+
+	"github.com/pingcap/tidb/pkg/types"
+	"github.com/pingcap/tidb/pkg/util/spill"
+)
+
+type SpillSerializeHelper struct {
+	// tmpBuf is an auxiliary data struct that used for encoding bytes.
+	// 1024 is large enough for all fixed length data struct.
+	tmpBuf [1024]byte
+
+	// varBuf is used for storing variable length data struct
+	varBuf []byte
+}
+
+func (s *SpillSerializeHelper) serializePartialResult4Count(value partialResult4Count) []byte {
+	return spill.SerializeInt64(int64(value), s.tmpBuf[0:int64Len])
+}
+
+func (s *SpillSerializeHelper) serializePartialResult4MaxMinInt(value partialResult4MaxMinInt) []byte {
+	spill.SerializeInt64(value.val, s.tmpBuf[0:int64Len])
+	end := int64Len + boolLen
+	spill.SerializeBool(value.isNull, s.tmpBuf[int64Len:end])
+	return s.tmpBuf[0:end]
+}
+
+func (s *SpillSerializeHelper) serializePartialResult4MaxMinUint(value partialResult4MaxMinUint) []byte {
+	spill.SerializeUint64(value.val, s.tmpBuf[0:uint64Len])
+	end := uint64Len + boolLen
+	spill.SerializeBool(value.isNull, s.tmpBuf[uint64Len:end])
+	return s.tmpBuf[0:end]
+}
+
+func (s *SpillSerializeHelper) serializePartialResult4MaxMinDecimal(value partialResult4MaxMinDecimal) []byte {
+	*(*types.MyDecimal)(unsafe.Pointer(&s.tmpBuf[0])) = value.val
+	end := types.MyDecimalStructSize + boolLen
+	spill.SerializeBool(value.isNull, s.tmpBuf[types.MyDecimalStructSize:end])
+	return s.tmpBuf[0:end]
+}
+
+func (s *SpillSerializeHelper) serializePartialResult4MaxMinFloat32(value partialResult4MaxMinFloat32) []byte {
+	spill.SerializeFloat32(value.val, s.tmpBuf[0:float32Len])
+	end := float32Len + boolLen
+	spill.SerializeBool(value.isNull, s.tmpBuf[float32Len:end])
+	return s.tmpBuf[0:end]
+}
+
+func (s *SpillSerializeHelper) serializePartialResult4MaxMinFloat64(value partialResult4MaxMinFloat64) []byte {
+	spill.SerializeFloat64(value.val, s.tmpBuf[0:float64Len])
+	end := float64Len + boolLen
+	spill.SerializeBool(value.isNull, s.tmpBuf[float64Len:end])
+	return s.tmpBuf[0:end]
+}
+
+func (s *SpillSerializeHelper) serializePartialResult4MaxMinTime(value partialResult4MaxMinTime) []byte {
+	*(*types.Time)(unsafe.Pointer(&s.tmpBuf[0])) = value.val
+	end := timeLen + boolLen
+	spill.SerializeBool(value.isNull, s.tmpBuf[timeLen:end])
+	return s.tmpBuf[0:end]
+}
+
+func (s *SpillSerializeHelper) serializePartialResult4MaxMinDuration(value partialResult4MaxMinDuration) []byte {
+	spill.SerializeInt64(int64(value.val.Duration), s.tmpBuf[0:int64Len])
+	spill.SerializeInt(value.val.Fsp, s.tmpBuf[int64Len:int64Len+intLen])
+	end := int64Len + intLen + boolLen
+	spill.SerializeBool(value.isNull, s.tmpBuf[int64Len+intLen:end])
+	return s.tmpBuf[0:end]
+}
+
+func (s *SpillSerializeHelper) serializePartialResult4MaxMinString(value partialResult4MaxMinString) []byte {
+	spill.SerializeBool(value.isNull, s.tmpBuf[0:boolLen])
+	resBuf := s.tmpBuf[:1]
+	resBuf = append(resBuf, value.val...)
```

Comment:
Why not store string in varBuf?

Reply:
> Why not store string in varBuf?

okk, I have refactored the `SpillSerializeHelper`